src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala

```diff
@@ -273,22 +269,45 @@ class WalletRegistry(store: HybridLDBKVStore)(ws: WalletSettings) extends Scorex
   /**
     * Updates scans of a box stored in the wallet database,
     * puts the box into the database if it is not there
+    * removes the box from the database if its there and scanIds are empty
     *
-    * @param scanIds
+    * @param newScans
     * @param box
     * @return
     */
-  def updateScans(scanIds: Set[ScanId], box: ErgoBox): Try[Unit] = Try {
+  def updateScans(newScans: Set[ScanId], box: ErgoBox): Try[Unit] = Try {
     val bag0 = KeyValuePairsBag(toInsert = Seq.empty, toRemove = Seq.empty)
-    val (updTb, bag1) = getBox(box.id) match {
-      case Some(tb) =>
-        (tb.copy(scans = scanIds), removeBox(bag0, tb))
-      case None =>
-        (TrackedBox(box, box.creationHeight, scanIds), bag0)
+    val oldBox = getBox(box.id)
+    val oldScans = oldBox.map(_.scans).getOrElse(Set.empty)
+
+    val newBox = TrackedBox(box, box.creationHeight, newScans)
+
+    val bag1 = (oldScans.isEmpty, newScans.isEmpty) match {
+      case (false, false) =>
+        putBox(removeBox(bag0, oldBox.get), newBox)
+      case (false, true) =>
+        removeBox(bag0, oldBox.get)
+      case (true, false) =>
+        putBox(bag0, newBox)
+      case (true, true) =>
+        throw new Exception("Can't remove a box which does not exist")
     }
-    val bag2 = putBox(bag1, updTb)
-    store.nonVersionedRemove(bag2.toRemove)
-    store.nonVersionedPut(bag2.toInsert)
+
+    val bag2 = if ((oldScans.contains(Constants.PaymentsScanId) || newScans.contains(Constants.PaymentsScanId)) &&
+      !(oldScans.contains(Constants.PaymentsScanId) && newScans.contains(Constants.PaymentsScanId))) {
```

Comment:
Please leave a clarifying comment on the meaning of the condition.

Reply:
Made a comment on an extracted digestChanged value